examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php

```diff
@@ -0,0 +1,602 @@
+     *     names
+     * @return MutateOperation[] the list of MutateOperations to remove all listing groups
+     */
+    // [START add_performance_max_product_listing_group_tree_3]
+    private static function createMutateOperationsForRemovingDescendents(
+        string $assetGroupListingGroupFilterResourceName,
+        array $parentsToChildren
+    ): array {
+        $operations = [];
+        if (array_key_exists($assetGroupListingGroupFilterResourceName, $parentsToChildren)) {
+            foreach ($parentsToChildren[$assetGroupListingGroupFilterResourceName] as $child) {
+                $operations = array_merge(
+                    $operations,
+                    self::createMutateOperationsForRemovingDescendents($child, $parentsToChildren)
+                );
+            }
+        }
+
+        $operations[] = new MutateOperation([
+            'asset_group_listing_group_filter_operation'
+                => new AssetGroupListingGroupFilterOperation([
+                    'remove' => $assetGroupListingGroupFilterResourceName
+                ])
+        ]);
+        return $operations;
+    }
+    // [END add_performance_max_product_listing_group_tree_3]
+
+    /**
+     * Creates a mutate operation that creates a root asset group listing group filter for the
+     * factory's asset group.
+     *
+     * The root node or partition is the default, which is displayed as "All Products".
+     *
+     * @param int $customerId the customer ID
+     * @param int $assetGroupId the asset group ID
+     * @param int $rootListingGroupId the root listing group ID
+     * @return MutateOperation the mutate operation for creating the root
+     */
+    // [START add_performance_max_product_listing_group_tree_4]
+    private static function createMutateOperationForRoot(
+        int $customerId,
+        int $assetGroupId,
+        int $rootListingGroupId
+    ): MutateOperation {
+        return self::createMutateOperationForSubdivision(
+            $customerId,
+            $assetGroupId,
+            $rootListingGroupId
+        );
+    }
+    // [END add_performance_max_product_listing_group_tree_4]
+
+    /**
+     * Creates a mutate operation that creates a intermediate asset group listing group filter.
+     *
+     * @param int $customerId the customer ID
+     * @param int $assetGroupId the asset group ID
+     * @param int $assetGroupListingGroupFilterId the ID of the asset group listing group filter to
+     *     be created
+     * @param int|null $parentId the ID of the parent of asset group listing group filter to be
+     *      created
+     * @param ListingGroupFilterDimension|null $listingGroupFilterDimension the listing group
+     *     filter dimension to associate with the asset group listing group filter
+     * @return MutateOperation the mutate operation for creating a subdivision
+     */
+    // [START add_performance_max_product_listing_group_tree_5]
```

Comment:
ditto: The golden includes the function comment (not sure if intended but let's keep it consistent)